rss2email module: init

Ekleog opened the pull request:
Also adding `system-sendmail` package for sharing the code with other
modules or packages needing it.

The test for this module depends on #48901 being merged, as the configuration format of OpenSMTPD changed. Hence the fact it's based on the tip of #48901.

cc @Profpatsch 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ekleog committed:
rss2email module: init

Also adding `system-sendmail` package for sharing the code with other
modules or packages needing it.

Ekleog commented:
@GrahamcOfBorg test rss2email

GrahamcOfBorg commented:
<!--REQUEST_ID=f9f2d7a3-a8a5-4737-9b04-cdf11f9ae094-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=bd9a7b23-9ac9-45a5-bbcb-f44bc75a0013)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 15.85s
cleaning up
killing server (pid 631)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/2nx6rhcksi6ccc9dyzfsnsjkqcsrqwl2-vm-test-run-opensmtpd
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=f9f2d7a3-a8a5-4737-9b04-cdf11f9ae094-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=a4d2eae2-dd33-44c1-aa54-699b7a082a79)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: running command: systemctl --no-pager show "rss2email"
server: exit status 0
error: unit ‘rss2email’ is inactive and there are no pending jobs
unit ‘rss2email’ is inactive and there are no pending jobs
cleaning up
killing server (pid 597)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
builder for '/nix/store/kypkz206qi0lgq24x95aavbjbdqmikk1-vm-test-run-opensmtpd.drv' failed with exit code 255
error: build of '/nix/store/kypkz206qi0lgq24x95aavbjbdqmikk1-vm-test-run-opensmtpd.drv' failed
```
</p></details>

Ekleog commented:
Looks like my `sleep` was optimistic despite being pessimistic… let's try again with something hopefully more resilient.

@GrahamcOfBorg test rss2email

GrahamcOfBorg commented:
<!--REQUEST_ID=8d229609-e60e-4b9d-9b57-fa623aa516e9-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=08f0ea7d-891b-490e-8b62-a2d6591953c0)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 14.75s
cleaning up
killing server (pid 631)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/wqya4ypjxnxlq8y0mirbx15l79adc50y-vm-test-run-opensmtpd
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=8d229609-e60e-4b9d-9b57-fa623aa516e9-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=f35babef-ccea-4575-9d7b-dc37938dea6b)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 19.56s
cleaning up
killing server (pid 597)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/q22xxxa68p9hczzs3q1dmwb0is9g5fjy-vm-test-run-opensmtpd
```
</p></details>

Profpatsch commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,221 @@
+        description = "How often to check the feeds, in systemd interval format";
+      };
+
+      configuration = mkOption {
```
You could use https://github.com/NixOS/nixpkgs/blob/master/lib/generators.nix and take an attrset instead.

Ekleog commented:
Thanks! I have applied your suggestion and rebased on master (as the opensmtpd update has been merged).

This should be ready to merge :)

@GrahamcOfBorg test rss2email

GrahamcOfBorg commented:
<!--REQUEST_ID=91a07ee9-1e89-4895-ba1a-38d53c241c3e-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=33378f9b-9973-4fc0-ae87-38433f15a75a)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 26.00s
cleaning up
killing server (pid 631)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/z47369agysjw8n2shahvzr9blq46vhr2-vm-test-run-opensmtpd
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=91a07ee9-1e89-4895-ba1a-38d53c241c3e-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=3febbb0c-62bc-493c-aa42-df8d590dc2b3)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 185.39s
cleaning up
killing server (pid 600)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/d8izh96fa61d5qysdx2bqqh0zkg5fgg5-vm-test-run-opensmtpd
```
</p></details>

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,226 @@
+        ExecStart =
+          "${cfg.package}/bin/r2e -c /var/rss2email/conf.cfg -d /var/rss2email/db.json run";
+        User = "rss2email";
+        PermissionsStartOnly = "true";
```
I think this would be a good usecase for the `RuntimeDirectory` and `DynamicUser` systemd options. Take a look at them (`man 5 systemd.exec`). And if the `db.json` file is state, you can probably use `StateDirectory`. Would simplify this service a lot.

Ekleog replied:
I thought of `DynamicUser` before rejecting it due to `db.json`… at the time I didn't know about `StateDirectory`, now I do, thanks! :)

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,226 @@
+        description = "How often to check the feeds, in systemd interval format";
+      };
+
+      configuration = mkOption {
```
I think it's more common to use the option name `config` for this.

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,226 @@
+        default = pkgs.rss2email;
+        defaultText = "pkgs.rss2email";
+        description = "The rss2email package to use.";
+      };
```
Is there a need to have this configurable? Is there an alternate package? If not, it's best to not provide this option because nobody will use it. And in any case, overlays can still change the value used by this module.

Ekleog replied:
There is no alternate package for the time being. However, I know I would personally use an alternate package (unreleased version), and I went on the `services.opensmtpd.package` precedent.

Also, I personally think that overlays are not really pleasant to use, and NUR packages are a much cleaner way of handling package changes, but a NUR package would require a local overlay to set it up with this module if the option was absent, so…

Infinisil replied:
I'd rather introduce this option when an actual need for it arises then. Upon a quick count, there's about 600 service modules in NixOS, and about 120 of them have a `package` option. If all of them would also introduce such an option, this would slow down evaluation quite a bit.

The 3 uses I see for a package option are:
- There are alternate packages you would want to use and being able to use different packages is an integral part of the module. Having an option makes this possibility discoverable. `postgresql` is such an example
- Using an overlay for it would require a lot of rebuilds, because overlays not only change the package the service uses, but all the other parts of the system as well. `curl` would be such an example (don't know of an example for a service though).
- Using an overlay for it would change some other part of the system you don't want to change. E.g. 2 services that need the same package but different versions of it. An overlay couldn't change just one of them.

Ekleog replied:
Well… I'm not of the same opinion, thinking that 600 options is basically nothing compared to the one-option-per-upstream-option vs. one-option-per-upstream-file, but let's agree to disagree :) I'm removing this option, as the usual behavior in NixOS appears to be not to have the `package` option.

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,226 @@
+            ''';
+          }
+        '';
+        default = default-config;
```
The problem with this is that setting *any* value in this option will remove all defaults. Also: is it really required to set these defaults in the NixOS config? If the program sets defaults itself if it can't find the values, then this isn't needed (also has the advantage that we don't need to update the defaults here if they change). If the defaults really do need to be set in order for it to work, I suggest doing this:
- Change the type to `attrsOf (either str (either int bool))`, this makes merging work as one would expect. When generating the result, make sure ints and bools get translated correctly (maybe the toINI generator does that already though).
- Set the defaults by doing `services.rss2email.config = mapAttrs (n: mkDefault) { use-8bit = false; ... };` in the `config` section of this module. This will make it such that all defaults aren't lost if a single gets changed. This also allows the user to see the value of the defaults by looking at the value of this option.

Ekleog replied:
I can confirm that the program uses its defaults internally without issues, thanks! Didn't think of this.

BTW, the merging-with-default behavior was handled before with `DEFAULT = default-config // cfg.configuration // { to = cfg.to; };`, which took `default-config`.

I'm still fixing the type as you advised so that the user can define parts of the option at different places. :)

I'm taking `default-config` and making it into a `literalExample`, so that the user has an idea of what they can use :)

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,226 @@
+
+            to = mkOption {
+              type = with types; nullOr str;
+              default = null;
```
Along with the `config` change, you can then make these changes here:
- Change the type to `str`
- Remove the default value here
- Set the default by doing `to = mkDefault cfg.to;` in the `config` section of the submodule.

Ekleog replied:
This is possible indeed, but it moves the module farther away from upstream configuration, as upstream configuration just doesn't add a `to`, and I tried to replicate upstream behavior as well as possible… now you're right, there's little else that could legitimately be done other than using the system-wide `to`, so I guess this won't change visible behavior :)

Infinisil replied:
Ah, I think your current implementation always just defaults to `cfg.to` though (line 12). So if you want to replicate upstream behavior, you should probably just not set it there instead.

Ekleog replied:
I think [lines 11-12](https://github.com/NixOS/nixpkgs/blob/829c58e2dec794347bac63faabf6f3f8819dcf17/nixos/modules/services/mail/rss2email.nix#L11-L12) correctly set `url` from `feed` and add the `to` field iff. `cfg.to` is not `null`, thanks to the `optionalAttrs` (and testing validates this, if I'm not mistaken on the testing process).

I guess you confused with [line 9](https://github.com/NixOS/nixpkgs/blob/829c58e2dec794347bac63faabf6f3f8819dcf17/nixos/modules/services/mail/rss2email.nix#L9), where I did set `to` unconditionally, but in the `[DEFAULT]` block?

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,226 @@
+
+  cfg = config.services.rss2email;
+  conf = pkgs.writeText "rss2email.cfg" (lib.generators.toINI {} ({
+      DEFAULT = default-config // cfg.configuration // { to = cfg.to; };
```
Along with the other changes, you can then set the `to` value via `services.rss2email.config.to = cfg.to` in the `config` section of this module.

Infinisil commented on pkgs/servers/mail/system-sendmail/default.nix:
```diff
@@ -0,0 +1,16 @@
+    echo "Unable to find system sendmail." >&2
+    exit 1
+  fi
+''
```
This is very ugly and I'd rather not have something like this in nixpkgs, but I don't know of a better solution right now.

Ekleog replied:
I agree, but already had to introduce it for `mdadm` in https://github.com/NixOS/nixpkgs/pull/38067 (OpenSMTPD's `sendmail` doesn't require setuid like postfix's), and putting it in a separate package will allow to share code with it (see https://github.com/NixOS/nixpkgs/pull/49229 that I'll rebase once this gets finished, and probably others I don't already know about because I haven't tested them on OpenSMTPD yet and they support only postfix because they have the security wrapper path hard-coded).

Thank you for your review! I'll review your other points when getting back to my computer where I can test things :)

Ekleog commented:
@GrahamcOfBorg test rss2email

GrahamcOfBorg commented:
<!--REQUEST_ID=1ef50742-b5f4-4047-8cc0-6c348f675722-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=bb8f4cb2-427e-4ca6-89d3-26fc44a1a519)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 24.20s
cleaning up
killing server (pid 597)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/g707rgchd4n74dxkmk5d605rm6s5n4nd-vm-test-run-opensmtpd
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=1ef50742-b5f4-4047-8cc0-6c348f675722-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49228&attempt_id=3d80f199-a082-4601-8a81-d85398467556)

Attempted: tests.rss2email

<details><summary>Partial log (click to expand)</summary><p>

```
server: exit status 1
syncing
server: running command: sync
server: exit status 0
test script finished in 22.29s
cleaning up
killing server (pid 631)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/ilhpsdz8yn6a2lkwi9ywg50cy7hrxi49-vm-test-run-opensmtpd
```
</p></details>

Ekleog commented:
Ugh hold. Apparently the test passes but on my server it fails due to `KeyError: 'getpwuid(): uid not found: 65140'`, so I guess actually `DynamicUser` isn't going to be a possibility… reverting this change and pushing again.

Ekleog commented:
Full traceback:
```
 r2e[18166]: Traceback (most recent call last):
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/site.py", line 544, in <module>
 r2e[18166]:     main()
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/site.py", line 530, in main
 r2e[18166]:     known_paths = addusersitepackages(known_paths)
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/site.py", line 282, in addusersitepackages
 r2e[18166]:     user_site = getusersitepackages()
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/site.py", line 258, in getusersitepackages
 r2e[18166]:     user_base = getuserbase() # this will also set USER_BASE
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/site.py", line 248, in getuserbase
 r2e[18166]:     USER_BASE = get_config_var('userbase')
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/sysconfig.py", line 601, in get_config_var
 r2e[18166]:     return get_config_vars().get(name)
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/sysconfig.py", line 558, in get_config_vars
 r2e[18166]:     _CONFIG_VARS['userbase'] = _getuserbase()
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/sysconfig.py", line 205, in _getuserbase
 r2e[18166]:     return joinuser("~", ".local")
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/sysconfig.py", line 184, in joinuser
 r2e[18166]:     return os.path.expanduser(os.path.join(*args))
 r2e[18166]:   File "/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/lib/python3.6/posixpath.py", line 249, in expanduser
 r2e[18166]:     userhome = pwd.getpwuid(os.getuid()).pw_dir
 r2e[18166]: KeyError: 'getpwuid(): uid not found: 65140'
```

Ekleog commented:
OK, can confirm that without `DynamicUser` again it works now, both on my server and in tests.

Ekleog commented:
It looks like @bjornfor's comment as well as my answer vanished on the `push --force` required for handling the merge conflict that arose from another package adding its id to `ids.nix`.

I reproduce them here for convenience: ([source](https://github.com/NixOS/nixpkgs/commit/29d7b28ac7d66f67abaa6d2d497c7c431b97a2b2#diff-d6445ebce1854ca67a03fc47fdb43c90R6))

(on the `if which sendmail > /dev/null 2>&1; then` line of `system-sendmail`)
@bjornfor 
> If you change which to command -v you remove one implicit dependency. (command -v is part of posix sh, whereas which is an external binary thay may or may not exist on the system.)

@Ekleog 
> Good point! Actually this mistake of mine hid a bug in the previous system-sendmail: when it's in $PATH then it infinite-looped.
>
> Both issues are fixed in latest push -f (required due to a merge conflict)

Infinisil commented on nixos/modules/services/mail/rss2email.nix:
```diff
@@ -0,0 +1,214 @@
+            imap-port = 143;
+            imap-ssl = false;
+            imap-mailbox = "INBOX";
+            verbose = "warning";
```
Now that we don't encode the default in the module anymore, I think this here should be removed as well and potentially link to some upstream default docs if available.

Ekleog replied:
Indeed, and actually these are in the manpage… or should be, if rss2email had had a release in the last 3 years. Well, with @Profpatsch we now have access to rss2email, so should be able to cleanup and make a release as soon as we get some time :)

I've pointed to the online `r2e.1` for the time being.

oxij commented:
`system-sendmail` has multiple lines in `meta.description`, I think this is generally frowned upon.

Also, as a side note, I think it would be good if somebody would finally add MDA support to `rss2email`. It is kinda crazy to run the whole of Postfix just to save files to users' maildirs. (I planned to do it 4 years ago...)

oxij commented:
Looking at it some more:

- I think

-  system-sendmail = callPackage ../servers/mail/system-sendmail { };
+  system-sendmail = lowPrio (callPackage ../servers/mail/system-sendmail { });

and

-    platforms = platforms.linux;
+    platforms = platforms.unix;

- Also, about defaulting to `1h`. Would it recheck all feeds every hour or would it respect recheck time supplied by the feeds themselves? It would be unfortunate if the former. In any case, I suggest setting it to at least `12h` by default.

Ekleog commented:
Jan Malakhovski <notifications@github.com> writes:

> Looking at it some more:
>
> - I think
>
> -  system-sendmail = callPackage ../servers/mail/system-sendmail { };
> +  system-sendmail = lowPrio (callPackage ../servers/mail/system-sendmail { });

`system-sendmail` will be a dependency of `mdadm` (once
https://github.com/NixOS/nixpkgs/pull/49229 will be merged), and it's a
script that's literally just writing a file, so I'm not sure it really
deserves being `lowPrio`.

Also, from `rg sendmail pkgs`, there seem to be quite a few packages
that would benefit from using it instead of hardcoding postfix's
security wrapper path… I guess no one ever tried to use them with a
decent smtpd that doesn't rely on setuid when it doesn't need to :p

> and
>
> -    platforms = platforms.linux;
> +    platforms = platforms.unix;

Done

> - Also, about defaulting to `1h`. Would it recheck all feeds every hour or would it respect recheck time supplied by the feeds themselves? It would be unfortunate if the former. In any case, I suggest setting it to at least `12h` by default.

Honestly, I have no idea. This should be raised on [1], I guess.

Raising to 12hr sounds good to me, I had literally no idea what to put
in there by default :)

[1] https://github.com/rss2email/rss2email/issues


I've pushed the updated version with the `linux` -> `unix` and
`1h` -> `12h` changes. Thank you for your review!

oxij commented:
>> -  system-sendmail = callPackage ../servers/mail/system-sendmail { };
>> +  system-sendmail = lowPrio (callPackage ../servers/mail/system-sendmail { });
>
> `system-sendmail` will be a dependency of `mdadm` (once
> https://github.com/NixOS/nixpkgs/pull/49229 will be merged), and it's a
> script that's literally just writing a file, so I'm not sure it really
> deserves being `lowPrio`.

`lowPrio` AFAIU lowers the priority for `buildEnv` and such, i.e. with `lowPrio` it should not override other `sendmail`s if you install several of them into `systemPackages`, which seems like a good idea to me.

> Also, from `rg sendmail pkgs`, there seem to be quite a few packages
> that would benefit from using it instead of hardcoding postfix's
> security wrapper path… I guess no one ever tried to use them with a
> decent smtpd that doesn't rely on setuid when it doesn't need to :p

Yes, I agree, I also grepped over `nixos` folder and got a bit discouraged by the amount of `sendmail` hackery there. Looking at the first few I wanted to propose to fix them too, but then it turned out there's a lot stuff to fix there, so I added it to my own low-prio TODO in case I get bored sometime.

>> - Also, about defaulting to `1h`. Would it recheck all feeds every hour or would it respect recheck time supplied by the feeds themselves? It would be unfortunate if the former. In any case, I suggest setting it to at least `12h` by default.
>
> Honestly, I have no idea. This should be raised on [1], I guess.
>
> Raising to 12hr sounds good to me, I had literally no idea what to put in there by default :)

I looked at the source and it does the wrong thing. IMHO, ideally, `rss2email` should just run as a daemon and handle all timeouts itself since there's no single timeout that works well for all RSS feeds (added to TODO). But given current state of `rss2email` the `12h` is the sanest thing that can be done for now, IMHO.

Ekleog commented:
Jan Malakhovski <notifications@github.com> writes:
>>> -  system-sendmail = callPackage ../servers/mail/system-sendmail { };
>>> +  system-sendmail = lowPrio (callPackage ../servers/mail/system-sendmail { });
>>
>> `system-sendmail` will be a dependency of `mdadm` (once
>> https://github.com/NixOS/nixpkgs/pull/49229 will be merged), and it's a
>> script that's literally just writing a file, so I'm not sure it really
>> deserves being `lowPrio`.
>
> `lowPrio` AFAIU lowers the priority for `buildEnv` and such, i.e. with `lowPrio` it should not override other `sendmail`s if you install several of them into `systemPackages`, which seems like a good idea to me.

Oh indeed I'm stupid, was thinking that was for lowering the priority of
the hydra job… which would make no sense. And say I once knew it :( too
many things.

Fixed :)

>> Also, from `rg sendmail pkgs`, there seem to be quite a few packages
>> that would benefit from using it instead of hardcoding postfix's
>> security wrapper path… I guess no one ever tried to use them with a
>> decent smtpd that doesn't rely on setuid when it doesn't need to :p
>
> Yes, I agree, I also grepped over `nixos` folder and got a bit discouraged by the amount of `sendmail` hackery there. Looking at the first few I wanted to propose to fix them too, but then it turned out there's a lot stuff to fix there, so I added it to my own low-prio TODO in case I get bored sometime.

Yeah, overall I'm thinking `system-sendmail` will be a good “factored”
hackery to at least make everything agree on the behavior. The issue
overall is that some packages need to depend on `sendmail`, and
`sendmail` is by essence impure, as it depends on the mailer
installed. NixOS could handle this better, but Nixpkgs in itself can't.

And as it can even be provided by things that relay to external mail
servers (eg. `msmtp`), we can't even just rely on a `sendmail` that
pipes the mail to `localhost:25`.

Maybe some hacks could be done with `/nix/var/run/sendmail-pipe` or
similar stuff, but…  let's say I'm not sure it'd actually be helpful.

>>> - Also, about defaulting to `1h`. Would it recheck all feeds every hour or would it respect recheck time supplied by the feeds themselves? It would be unfortunate if the former. In any case, I suggest setting it to at least `12h` by default.
>>
>> Honestly, I have no idea. This should be raised on [1], I guess.
>>
>> Raising to 12hr sounds good to me, I had literally no idea what to put in there by default :)
>
> I looked at the source and it does the wrong thing. IMHO, ideally, `rss2email` should just run as a daemon and handle all timeouts itself since there's no single timeout that works well for all RSS feeds (added to TODO). But given current state of `rss2email` the `12h` is the sanest thing that can be done for now, IMHO.

At that point I'd guess you'd maybe do it better by just writing a
`rss2email` daemon from scratch, I'm not sure you could re-use lots of
current `rss2email` apart from its RSS-parsing library and its
mail-sending library?

Anyway, this is getting kind of off-topic :)


I think all the concerns are now solved, and there appear not to be
merge conflicts currently.

7c6f434c commented:
(triage) @Infinisil are all your concerns addressed?

Infinisil commented:
@7c6f434c Yeah they're addressed. I still don't like the wrapper thing, but I guess if that's the only way then sure.

7c6f434c commented:
I see why you don't like the need for the wrapper, but I also don't see a clean and cheap way around that.
